### ISSUE TITLE
ENH: stats: adding more random number generators from UNU.RAN

### ISSUE THREAD

**Issue by tirthasheshpatel** [scipy/scipy#14600]
UNU.RAN contains many generators to sample from univariate continuous and discrete distributions. gh-14215 adds two such generators `TransformedDensityRejection` and `DiscreteAliasUrn`. Over time it would be great to add these generators to SciPy:

For continuous distributions:

- [ ] `NumericalInversePolynomial`
- [ ] `NumericalInversion`
- [ ] `AutomaticRatioOfUniforms`
- [ ] `NaiveRatioOfUniforms`
- [ ] `SimpleRatioOfUniforms`
- [ ] `AdaptiveRejectionSampling`
- [ ] `InverseTransformedDensityRejection`
- [ ] `SimpleSetupRejection`
- [x] `TransformedDensityRejection`
- [ ] `UniversalTransformedDensityRejection`
- [ ] `PiecewiseConstantHatsTable`

For discrete distributions:

- [ ] `DiscreteAutomaticRejectionInversion`
- [x] `DiscreteAliasUrn`
- [ ] `DiscreteGuideTable`
- [ ] `DiscreteSimpleRatioOfUniforms`
- [ ] `DiscreteSequentialSearch`

More details of all these generators can be found in [this spreadsheet](https://docs.google.com/spreadsheets/d/1r36HypXwpit7sHt9YAe3K7yPL7HNNuQjTkbVYQIJ4xI/edit?usp=sharing).

**Comment by tupui** [comment]
Thanks for this list. There are interesting things.

Do we want all of it? or you listed all possibilities? I am thinking about `NumericalInversion` for instance which we might not need as we already have some `ppf` and `NumericalInverseHermite`.

As for how to add these, is there a "recipe" to follow?

**Comment by tirthasheshpatel** [comment]
> Do we want all of it? or you listed all possibilities? I am thinking about `NumericalInversion` for instance which we might not need as we already have some `ppf` and `NumericalInverseHermite`.

Right. We don't need to add all it: for example, we don't need `NumerialInverse` (the last time I tried it out, it was sometimes even slower than the default method in SciPy). I and @chrisb83 decided to prioritize the methods that are faster and widely applicable: for example, `NumerialInversePolynomial` and `AutomaticRatioOfUniforms`.

> As for how to add these, is there a "recipe" to follow?

I haven't written a developer guide yet but gh-14215 has two examples which can be followed to add more generators. They follow a format like this:

```python
cdef class GeneratorName(Method):
    def __cinit__(self, dist, center=None, ..., *, domain=None, ...):
        # save all keyword argument for pickling support
        self._kwargs = {'dist': dist, 'domain': domain, ...}
        cdef unur_distr *distr
        cdef unur_par *par
        cdef unur_gen *rng

        dist, domain, ... = self._validate_args(dist, domain, ...)

        # save callbacks, if any
        self.callbacks = _unpack_dist(dist, meths=['pdf'], optional_meths=['cdf', ...])
        def _callback_wrapper(x, name):
            return self.callbacks[name](x)
        self._callback_wrapper = _callback_wrapper

        # call all the UNU.RAN C functions under a lock
        self._messages = MessageStream()
        _lock.acquire()
        try:
            unur_set_stream(self._messages.handle)

            # initialize a distribution
            self.distr = unur_distr_cont_new()
            if self.distr == NULL:
                raise UNURANError(self._messages.get())
            _pack_distr(self.distr, self.callbacks)

            # set distribution methods, attributes, etc
            if center is not None:
                self._check_errorcode(unur_distr_cont_set_center(self.distr, center))
            ...

            if domain is not None:
                self._check_errorcode(unur_distr_cont_set_domain(self.distr, domain[0],
                                                                 domain[1]))

            # initialize a parameter object
            self.par = unur_pinv_new(self.distr)
            if self.par == NULL:
                raise UNURANError(self._messages.get())

            # set parameter attributes
            self._check_errorcode(unur_generator_set_whatever(self.par, whatever))
            ...

            # initialize a UNU.RAN generator
            self._set_rng(random_state)
        finally:
            _lock.release()

    cdef object _validate_args(dist, domain, ...):
        domain = _validate_domain(domain, dist)
        return dist, domain, ...

    # attributes
    ...

    # methods
    ...

```

**Comment by Kai-Striega** [comment]
This looks interesting and I'd be willing to help. I don't have much experience wrapping C functions but I'd be willing to learn given some guidance. @tirthasheshpatel @chrisb83 would you like some help with this? And would either of you be able to give me some guidance on where to go from here if I'm willing to help?

**Comment by chrisb83** [comment]
@Kai-Striega It would be great if you help us! 

gh-14717 should give a good overview of the required code changes to add a new method. Both Tirth and I noted that writing the docs / tutorial takes up most of the time (though one can copy quite a few things from UNU.RAN).

Tirth will soon work on SimpleRatioUniforms and I will implement AutomaticRatioUniforms. You can take a look at any of the other methods. I think DiscreteGuideTable would be nice to add (https://statmath.wu.ac.at/unuran/doc/unuran.html#DGT)

Let us know if you would like to have a virtual meeting to discuss this in more detail

**Comment by tirthasheshpatel** [comment]
> would you like some help with this?

Sure, @Kai-Striega!

> And would either of you be able to give me some guidance on where to go from here if I'm willing to help?

To get started, you should refer to UNU.RAN's user manual for the generator you wish to wrap. This should give you some idea about what distribution methods are required (e.g. `pdf`, `cdf`, etc) and the parameters for that generator. A while ago, I wrote [this spreadsheet](https://docs.google.com/spreadsheets/d/1r36HypXwpit7sHt9YAe3K7yPL7HNNuQjTkbVYQIJ4xI/edit?usp=sharing) to help with this but it's a bit outdated now and best is to refer UNU.RAN docs. You can copy-paste UNU.RAN docs to write the API documentation. It is quite time-consuming to write tutorials so, if you have detailed API documentation, you can just copy-paste that into the tutorials for now. I think the rest can be handled in reviews.

As @chrisb83 said, I think adding `DiscreteGuideTable` would be very helpful as we only have one such generator to sample from discrete distributions.

If you want any examples, you can look at any of the related PRs (gh-14703, gh-14717).

> I don't have much experience wrapping C functions but I'd be willing to learn given some guidance.

I have tried to abstract out most C/Cython details required to wrap a generator. So, it should be mostly easy to contribute without knowing them well. If you do face any problem though, we can probably resolve it in reviews.

**Comment by Kai-Striega** [comment]
> It would be great if you help us! gh-14717 should give a good overview of the required code changes to add a new method. 

Great I'll get started with reading into it!

> Both Tirth and I noted that writing the docs / tutorial takes up most of the time (though one can copy quite a few things from UNU.RAN).

I'll start the wrapper first. I don't really mind writing so it's not too much of a problem for me.

> I think DiscreteGuideTable would be nice to add

I'll take a look at it

> Let us know if you would like to have a virtual meeting to discuss this in more detail

I think I'll try to get  a PR done first just from text for now. I'm going in and out of rural Australia atm so my internet connection will be interesting for a while. If it doesn't work out I'll get to Perth and we can have a meeting.

>  while ago, I wrote this spreadsheet to help with this but it's a bit outdated now and best is to refer UNU.RAN docs.

That looks useful. I'll start reading that and the docs. Chris linked them above so they're pretty easy to find.

> You can copy-paste UNU.RAN docs to write the API documentation. It is quite time-consuming to write tutorials so, if you have detailed API documentation, you can just copy-paste that into the tutorials for now. I think the rest can be handled in reviews.

:+1: 

> I have tried to abstract out most C/Cython details required to wrap a generator. So, it should be mostly easy to contribute without knowing them well. If you do face any problem though, we can probably resolve it in reviews.

That sounds reassuring. I'll get started but it might take me a while to get my PR in.

**Comment by Kai-Striega** [comment]
With ``DiscreteGuideTable`` merged I could, if it's reasonably similar, start working on ``SimpleRatioUniforms``.

**Comment by chrisb83** [comment]
I will close this issue since we do not plan to add further methods for now
